docs/api/js/utils.md

```diff
@@ -1,68 +1,785 @@
+}
+```
+
+See also [`undol1tol2alias`](#undol1tol2alias).
+
+### `checkBaseCost`
+
+Checks the base cost is within prescribed limits.
+
+#### Inputs
+
+- `baseCost`: base cost ?? as `BigNumber` object.
+- `value`: value ?? as `BigNumberIsh` object.
+
+#### Outputs
+
+- Returns an `Error` message if the limit is breached.
+
+```ts
+export async function checkBaseCost(baseCost: ethers.BigNumber, value: ethers.BigNumberish | Promise<ethers.BigNumberish>) {
+    if (baseCost.gt(await value)) {
+        throw new Error(
+            `The base cost of performing the priority operation is higher than the provided value parameter ` +
+                `for the transaction: baseCost: ${baseCost}, provided value: ${value}`
+        );
+    }
+}
+```
+
+### `create2Address`
+
+Generates a future-proof contract address using salt plus bytecode which allows determination of an address before deployment. 
+
+:::warning
+- The zkSync Era implementation is slightly different from Ethereum.
+:::
+
+#### Inputs
+
+- `sender`: sender address as string.
+- `bytecodeHash`: output from zkSolc as `BytesLike` object.
+- `salt`: randomization element as `BytesLike` object.
+- `input`: ABI encoded constructor arguments as `BytesLike` object.
+
+#### Outputs
+
+- Returns an `Address` object.
+
+```ts
+export function create2Address(sender: Address, bytecodeHash: BytesLike, salt: BytesLike, input: BytesLike) {
+    const prefix = ethers.utils.keccak256(ethers.utils.toUtf8Bytes('zksyncCreate2'));
+    const inputHash = ethers.utils.keccak256(input);
+    const addressBytes = ethers.utils
+        .keccak256(ethers.utils.concat([prefix, ethers.utils.zeroPad(sender, 32), salt, bytecodeHash, inputHash]))
+        .slice(26);
+    return ethers.utils.getAddress(addressBytes);
+}
+```
+
+:::tip
+The `prefix` in the `kekkak256` calculation is equal to `keccak256("zksyncCreate")`.
+:::
+
+### `createAddress`
+
+Generates a contract address from deployer's account and nonce.
+
+#### Inputs
+
+- `sender`: sender address as string.
+- `senderNonce`: sender nonce as `BigNumberish` object.
+
+#### Outputs
+
+- Returns an `Address` object.
+
+```ts
+export function createAddress(sender: Address, senderNonce: BigNumberish) {
+    const prefix = ethers.utils.keccak256(ethers.utils.toUtf8Bytes('zksyncCreate'));
+    const addressBytes = ethers.utils
+        .keccak256(
+            ethers.utils.concat([
+                prefix,
+                ethers.utils.zeroPad(sender, 32),
+                ethers.utils.zeroPad(ethers.utils.hexlify(senderNonce), 32)
+            ])
+        )
+        .slice(26);
+
+    return ethers.utils.getAddress(addressBytes);
+}
+```
+
+### `eip712TxHash`
+
+Returns signed keccak256 representation of a transaction with signature.
+
+#### Inputs
+
+- `transaction`: given by id ?? as ??
+- `ethSignature?`: the transaction signer signature ?? as `EthereumSignature` object (optional).
```

Comment:
optional ECDSA signature of the transaction

Reply:
done